File: query-daemon.js
#!/usr/bin/env node

const WebSocket = require('ws');
const crypto = require('crypto');

function query(queryType, timeout = 5000) {
  return new Promise((resolve, reject) => {
    const requestId = crypto.randomUUID();
    const ws = new WebSocket('ws://localhost:8082');
    
    const timer = setTimeout(() => {
      ws.close();
      reject(new Error('Query timeout'));
    }, timeout);
    
    ws.on('open', () => {
      ws.send(JSON.stringify({ id: requestId, query: queryType }));
    });
    
    ws.on('message', (data) => {
      try {
        const response = JSON.parse(data.toString());
        if (response.id === requestId) {
          clearTimeout(timer);
          ws.close();
          
          if (response.error) {
            reject(new Error(response.error));
          } else {
            resolve(response.result);
          }
        }
      } catch (e) {
        clearTimeout(timer);
        ws.close();
        reject(e);
      }
    });
    
    ws.on('error', (err) => {
      clearTimeout(timer);
      reject(err);
    });
  });
}

// CLI usage
if (require.main === module) {
  const queryType = process.argv[2] || 'canvas_size';
  
  query(queryType)
    .then(result => {
      console.log(JSON.stringify(result, null, 2));
      process.exit(0);
    })
    .catch(err => {
      console.error(`Error: ${err.message}`, { file: process.stderr });
      process.exit(1);
    });
}

module.exports = { query };

File: before 2026-02-15/query-daemon.py
#!/usr/bin/env python3

import asyncio
import websockets
import json
import sys
import uuid

async def query(query_type, host='localhost', port=8082, timeout=5.0):
    """
    Send a query to the daemon and wait for response.
    
    Args:
        query_type: Type of query ('canvas_size', 'defaults', 'objects')
        host: Daemon hostname
        port: Daemon WebSocket port
        timeout: Timeout in seconds
    
    Returns:
        Query result as Python dict/object
    """
    request_id = str(uuid.uuid4())
    query_msg = json.dumps({"id": request_id, "query": query_type})
    
    uri = f"ws://{host}:{port}"
    
    try:
        async with websockets.connect(uri) as websocket:
            await websocket.send(query_msg)
            
            response = await asyncio.wait_for(websocket.recv(), timeout=timeout)
            data = json.loads(response)
            
            if data.get('id') == request_id:
                if 'error' in data:
                    raise Exception(data['error'])
                return data.get('result')
            else:
                raise Exception('Response ID mismatch')
                
    except asyncio.TimeoutError:
        raise Exception('Query timeout')

# CLI usage
if __name__ == '__main__':
    query_type = sys.argv[1] if len(sys.argv) > 1 else 'canvas_size'
    
    try:
        result = asyncio.run(query(query_type))
        print(json.dumps(result, indent=2))
    except Exception as e:
        print(f'Error: {e}', file=sys.stderr)
        sys.exit(1)
